x/incentive/keeper/update.go

```diff
@@ -6,14 +6,194 @@ import (
 //
+// Unlike updateRewards and updatePrograms, this function is not called during EndBlock.
+//
 // REQUIREMENT: This function must be called during any message or hook which creates an unbonding or updates
 // bonded amounts. Leverage hooks which decrease borrower collateral must also call this before acting.
 // This ensures that between any two consecutive claims by a single account, bonded amounts were constant
-// on that account for each bond tier and collateral uToken denom.
-func (k Keeper) updateAccount(_ sdk.Context, _ sdk.AccAddress) (sdk.Coins, error) {
-	// TODO
-	return sdk.Coins{}, incentive.ErrNotImplemented
+// on that account for each collateral uToken denom.
+func (k Keeper) UpdateAccount(ctx sdk.Context, addr sdk.AccAddress) (sdk.Coins, error) {
+	// unbondings have already been subtracted from bonded amounts when they are started,
+	// so it is fine to finish completed unbondings before claiming rewards.
+	if err := k.finishUnbondings(ctx, addr); err != nil {
+		return sdk.NewCoins(), err
+	}
+
+	return k.claimRewards(ctx, addr)
+}
+
+// updateRewards updates any rewardAccumulators associated with ongoing incentive programs
+// based on the time elapsed between LastRewardTime and block time. It also decreases active
+// incentive programs' RemainingRewards by the amount of coins distributed.
+func (k Keeper) updateRewards(ctx sdk.Context, prevTime, blockTime int64) error {
+	if prevTime > blockTime {
+		return incentive.ErrDecreaseLastRewardTime
+	}
+	if prevTime == blockTime {
+		return nil
+	}
+	timeElapsed := blockTime - prevTime
+	ongoingPrograms, err := k.getAllIncentivePrograms(ctx, incentive.ProgramStatusOngoing)
+	if err != nil {
+		return err
+	}
+	for _, p := range ongoingPrograms {
+		bondedDenom := p.UToken
+		bonded := k.getTotalBonded(ctx, bondedDenom)
+
+		// calculate the amount of time (in seconds) that remained on the incentive
+		// program after the previous calculation.
+		prevRemainingTime := (p.StartTime + p.Duration) - prevTime
+		// if remaining time was zero or negative, but program had not been removed
+		// from ongoing programs, using a value of 1 second ensures all its remaining
+		// rewards are distributed
+		if prevRemainingTime < 1 {
+			prevRemainingTime = 1
+		}
+
+		// The portion of a program's remaining rewards,
+		// ranging from 0 to 1, which will be distributed this
+		// block. The max value of 1 means 100% of remaining rewards
+		// will be used, as occurs when a program is ending.
+		programRewardsFraction := sdk.MinDec(
+			sdk.OneDec(),
+			sdk.NewDecFromInt(sdk.NewInt(timeElapsed)).
+				Quo(sdk.NewDec(prevRemainingTime)))
+
+		// each incentive program has only one reward denom
+		rewardDenom := p.RemainingRewards.Denom
+
+		// get this block's rewards (as a token amount) for this incentive program only
+		thisBlockRewards := sdk.NewCoin(
+			rewardDenom,
+			sdk.NewDecFromInt(p.RemainingRewards.Amount).Mul(programRewardsFraction).TruncateInt())
```

Comment:
Current math is more adaptive in case of significant rounding error - where this would delay all rounded down rewards until the last block, current approach distributes them more smoothly.

I will doublecheck the math, but probably leaving unchanged.

Reply:
let's create a task to simulate the diff (eg in python) and compare

Reply:
https://linear.app/umee/issue/UMEE-578/incentives-simplify-reward-distribution